src/lib/cache/codec.go

```diff
@@ -0,0 +1,76 @@
+// Copyright Project Harbor Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cache
+
+import (
+	"sync"
+
+	"github.com/goharbor/harbor/src/lib/log"
+	"github.com/vmihailenco/msgpack/v5"
+)
+
+// Codec codec interface for cache
+type Codec interface {
```

Comment:
I'm not sure if this is a must have for now.

Additionally, IMO the codec should be used by the reader/writer, for example, if we decide to store all objects into a cache, different objects may be encoded/decoded differently?

Reply:
`Codec` is introduced to reduce the cost to use the cache library. We don't need to handle the details of how to save a struct to the cache, the fetched value from the cache is the type we wanted.

Without the codec implicitly related to the cache, we need to decode the value to the type the cache accepted before save it to the cache, usually, the accepted type is a string or bytes. And we also need to encode the value to the type we wanted after the value fetched from the cache.

The only problem for the codec is that it needs to deal with most types, luckily the msgpack works with primitives, arrays, maps, structs, time.Time and interface{}.